tket/src/Architecture/Architecture.hpp

```diff
@@ -97,27 +131,33 @@ class Architecture : public graphs::UIDConnectivity<Node> {
 JSON_DECL(Architecture::Connection)
 JSON_DECL(Architecture)
 
-// Subclass, constructor generates adjacency matrix corresponding to a fully
-// connected architecture
-class FullyConnected : public Architecture {
+class FullyConnected : public ArchitectureBase<graphs::CompleteGraph<Node>> {
  public:
-  explicit FullyConnected(unsigned numberOfNodes);
-  // get_all_uids() does not guarantee to return nodes in any order
-  // this returns the canonical ordering of nodes
-  static node_vector_t get_nodes_canonical_order(unsigned numberOfNodes);
+  FullyConnected() : ArchitectureBase<graphs::CompleteGraph<Node>>() {}
 
- private:
-  static std::vector<Connection> get_edges(unsigned numberOfNodes);
+  /**
+   * Construct a fully-connected graph of a given size.
+   *
+   * The nodes are labelled "fcNode" (indexed from 0 to n-1).
+   *
+   * @param n number of nodes
+   */
+  explicit FullyConnected(unsigned n) {
+    for (unsigned i = 0; i < n; i++) {
+      nodes_.insert(Node("fcNode", i));
+    }
+  }
 };
 
+JSON_DECL(FullyConnected)
+
 // Subclass, constructor generates adjacency matrix corresponding to a ring
 // architecture
 class RingArch : public Architecture {
```

Comment:
This could be inheriting from `ArchitectureBase` directly, without relying on an explicit BGL graph.

Reply:
Same for SquareGrid below.

Reply:
Would rather leave these here to avoid having to reimplement a bunch of methods for them, at least until there is some advantage to doing so.

Reply:
I was thinking that these architectures, just like `FullyConnected` don't need to keep the graph in memory because of their regular structure. But I guess that unlike `FullyConnected`, these are sparse, so it shouldn't be too much of a problem.